pkg/kubelet/pluginmanager/pluginwatcher/README.md

```diff
@@ -16,6 +16,87 @@ there.
+- The new instance starts, unlinks that file, and starts listening on it again.
+- In parallel, the kubelet notices the removal and unregisters the plugin
+  before probing the new instance, thus breaking the seamless upgrade.
+
+Even if the timing is more favorable and unregistration is avoided, using the
+same socket is problematic: if the new instance fails, the kubelet cannot fall
+back to the old instance because that old instance is not listening to the
+socket that is available under `plugin.example.com-reg.sock`.
+
+This can be achieved in a DaemonSet by passing the UID of the pod into the pod
+through the downward API. New instances may try to clean up stale sockets of
+older instances, but have to be absolutely sure that those sockets really
+aren't in use anymore. Each instance should catch termination signals and clean
+up after itself. Then sockets only leak during abnormal events (power loss,
+killing with SIGKILL).
+
+Last but not least, both plugin instances must be usable in parallel. It is not
+predictable which instance the kubelet will use for which request.
+
+### In the kubelet
+
+For such a seamless upgrade with different sockets per plugin to work reliably,
+the handler for the plugin type must track all registered instances. Then if
+one of them fails and gets unregistered, it can fall back to some
+other. Picking the most recently registered instance is a good heuristic. This
+isn't perfect because after a kubelet restart, plugin instances get registered
+in a random order. Restarting the kubelet in the middle of an upgrade should be
+rare.
+
+At the moment, the following handlers do not support such seamless upgrades:
+
+- The device plugin handler suffers from temporarily removing the extended
+  resources during an upgrade. A proposed fix is pending in
+  https://github.com/kubernetes/kubernetes/pull/127821.
+
+- The CSI handler [tries to determine which instance is newer](https://github.com/kubernetes/kubernetes/blob/7140b4910c6c1179c9778a7f3bb8037356febd58/pkg/volume/csi/csi_plugin.go#L115-L125) based on the supported version(s) and
+  only remembers that one. If that newest instance fails, there is no fallback.
+
+  In practice, most CSI drivers probably all pass [the hard-coded "1.0.0"](https://github.com/kubernetes-csi/node-driver-registrar/blob/27700e2962cd35b9f2336a156146181e5c75399e/cmd/csi-node-driver-registrar/main.go#L72)
+  from the csi-node-registrar as supported version, so this version
+  selection mechanism isn't used at all.
+
+This supports it:
+
+- DRA
+
+### Deployment
+
+Deploying a plugin with support for seamless upgrades and per-instance socket
+filenames is *not* compatible with a kubelet version that does not have support
+for seamless upgrades yet. It breaks like this:
+
+- New instance starts, gets registered and replaces the old one.
+- Old instance stops, removing its socket.
+- The kubelet notices that, unregisters the plugin.
```

Comment:
So, it looks possible that `PrepareResources` is called for the old instance and `UnprepareResources` - for the new one. How new instance would "know" that the resource is prepared when receiving `UnprepareResources` request?  Is this info stored in the slice and new instance should fetch it before processing the request?

Reply:
How to persist state across container restarts is a problem that DRA drivers must solve themselves. This PR doesn't change anything in that regard.

They can use local files for that or reconstruct the current state of the node on a container start. They cannot use anything in the API, there are no fields for that.

Reply:
> How to persist state across container restarts is a problem that DRA drivers must solve themselves. 

Is this mentioned anywhere in the documentation? DRA drivers developers should be informed about this.
BTW, this could be different from handling restarts as far as I understand. In this scenario it's possible that two drivers are running at the same time and not restarting. Kubelet will call PrepareResources for the old driver and UnprepareResources for another one and both should be able to handle this somehow.

Reply:
> Is this mentioned anywhere in the documentation? DRA drivers developers should be informed about this.

I'm pretty sure we mentioned it. The DRA example driver implements it.

> BTW, this could be different from handling restarts as far as I understand. In this scenario it's possible that two drivers are running at the same time and not restarting. Kubelet will call PrepareResources for the old driver and UnprepareResources for the new one and both should be able to handle this somehow.

The `RollingUpdate` calls out the "running two plugin instances in parallel" aspect. It's left to the DRA driver author to recognize what that means. Let me be more explicit...

Reply:
```
...
// There is no guarantee which of the two instances are used by kubelet.
// For example, it can happen that a claim gets prepared by one instance
// and then needs to be unprepared by the other. Kubelet then may fall back
// to the first one again for some other operation. In practice this means
// that each instance must be entirely stateless across method calls.
// Serialization (on by default, see [Serialize]) ensures that methods
// are serialized across all instances through file locking. The plugin
// implementation can load shared state from a file at the start
// of a call, execute and then store the updated shared state again.
...
func RollingUpgrade ...
```